### ISSUE TITLE
API to determine whether integrated terminal has focus

### ISSUE THREAD

**Issue by pokey** [microsoft/vscode#117980]
As part of implementing improved voice coding support, we are developing a VSCode extension that enables improved communication with [talon](https://talonvoice.com/).  One thing we need is to be able to determine that the terminal pane currently has focus, so that we can activate commands for interacting with the terminal.  To do so, we'd basically need two things:
- A function to check whether the terminal has focus
- An event that is fired when the terminal gains / loses focus

**Comment by pokey** [comment]
Fwiw I think half of this issue could be solved with https://github.com/microsoft/vscode/pull/118722 by using the `terminalFocus` context

**Comment by Tyriar** [comment]
Right now our API doesn't really have the concept of focus I think by design. We'll see how https://github.com/microsoft/vscode/pull/118722 turns out but I think the right thing to do here is to hook into the operating system's accessibility APIs and check what is being focused in the accessibility tree.

**Comment by pokey** [comment]
@TylerLeonhardt thanks for your response.  Not sure I understand, though.   Could you possibly explain why your design encourages exposing focus as a when-clause context, but not as an API?  Thanks for your help

**Comment by Tyriar** [comment]
This is what I was told:

> the API doesn’t really have the concept of focus, because focus if often in elements that have no API representation

**Comment by pokey** [comment]
Ok, well if your source has a response to my question, I'd be interested to understand the answer, but I guess hopefully

**Comment by daanzu** [comment]
@Tyriar Having this would be tremendously useful for accessibility. It is quite disappointing to hear that there is push back to implementing it.

**Comment by daanzu** [comment]
@pokey I hope such an extension could be made general enough to be used by voice software other than talon. I have been wanting to work on something similar for my own setup using [dragonfly/kaldi-active-grammar](https://github.com/daanzu/kaldi-active-grammar). Specifically support for: executing VSCode commands, retrieving VSCode context, and retrieving VSCode's buffer text around the cursor for priming the dictation language model. Is your work being done on github?

**Comment by pokey** [comment]
Hi @daanzu glad to hear the interest! None of my extensions are talon-specific. I have a few extensions that I've developed for voice coding, but you'll probably be most interested in https://marketplace.visualstudio.com/items?itemName=pokey.command-server. Lmk if you have any troubles getting it set up / linked with kaldi and I'd be happy to hop on a zoom and pair up!

**Comment by Tyriar** [comment]
Why can't you use the OS' a11y APIs to do this? I would have expected that's possible as it's how screen readers get all their information.

**Comment by daanzu** [comment]
@Tyriar It should be possible, but the OS accessibility API tends to be quite complicated, slow, and prone to breaking. If I am understanding it correctly, the VSCode API already has the concept for key bindings, which has the benefit of being easily discoverable for users. It seems silly to then make it unavailable as one simple API query.

@pokey Thanks for the info! I'll take a look.

**Comment by lunixbochs** [comment]
> Why can't you use the OS' a11y APIs to do this? I would have expected that's possible as it's how screen readers get all their information.

On Mac, VSCode appears to the accessibility subsystem by default as an opaque rectangle. In Electron, you can manually set a flag on the window to enable enhanced accessibility, but that doesn't work for everything, and when it does work it sort of exports the whole DOM which typically comes with a huge performance and resource hit (which is why it's disabled by default in Chromium and Electron - they aren't exactly light on resources to begin).

The simplest way to surface context information (e.g. which part of the app is active) to voice control systems is to put a simple string in the window title.

**Comment by sourishkrout** [comment]
> This is what I was told:
> 
> > the API doesn’t really have the concept of focus, because focus if often in elements that have no API representation

Would it alternatively be possible to fire `vscode.window.onDidChangeActiveTerminal` when the terminal comes into focus? That way one could imply `focus` if, let's say, a `vscode.window.onDidChangeActiveTextEditor` was fired last. Again, it's implicit but would be consistent with how `vscode.window.onDidChangeActiveTextEditor` works. This event fires again if the focus goes back to the same file; even works when both files are exposed side by side (E.g. view column 1 & 2).

My use-case requires me to reasonably triangulate if you're currently active in the editor or terminal. `focused` analogous to `vscode.window.onDidChangeWindowState` would be ideal, however, firing `vscode.window.onDidChangeActiveTerminal` every time the focus changes to active on a terminal would be sufficent.

Thoughts @pokey @Tyriar? Thanks!

**Comment by Tyriar** [comment]
@sourishkrout I think that blurs the lines a little too much better active and focus, right now it only fires when the actual active terminal is changed (last focused terminal instance within the active tab). I tested `vscode.window.onDidChangeActiveTextEditor` with the below and it doesn't fire on focus, just active like the terminal API?

```ts
vscode.window.onDidChangeActiveTextEditor(e => {
  vscode.window.showInformationMessage('focus text editor ' + e?.document.uri.fsPath);
});
```

**Comment by sourishkrout** [comment]
Hm. That's odd. I've used your code snippet and get below's result, @Tyriar. Fires every time. Does this match what you're seeing?

I'm open to other approaches: E.g. a property/method on `vscode.window.activeTerminal` or `vscode.window.terminals` - would that be an option? I'd really appreciate a path forward or a workaround.

![events](https://user-images.githubusercontent.com/250527/113429165-ee70b380-938c-11eb-90f3-5ffb00e16efc.gif)

**Comment by Tyriar** [comment]
But there the active editor is changing, just as if you had 2 terminals side by side the terminal API would do that as well?

**Comment by sourishkrout** [comment]
> But there the active editor is changing, just as if you had 2 terminals side by side the terminal API would do that as well?

That's fair. I do see how the active terminal events fires semantically works the same as active editor if I use split terminals. 
Thinking about alternatives (e.g. "how do i know the terminal's active?") I've stumbled over https://github.com/microsoft/vscode/issues/79063. Going to drop a comment on that.

**Comment by lunixbochs** [comment]
Bumping this thread for follow-up discussion because https://github.com/microsoft/vscode/pull/118722 was closed by @jrieken and the accessibility discussions there don't have a resolution or path forward yet.

https://github.com/microsoft/vscode/pull/118722#issuecomment-805911807 - why getContext isn't the answer
https://github.com/microsoft/vscode/pull/118722#issuecomment-806678493 - why VSCode can't completely own the when-clause

It's a huge limitation if accessibility tools must act as a one-way event source. We have large context-specific command sets that can change based on stuff like the current language, whether you're in a terminal, current terminal working directory, etc.

An example voice command use of this, (which wouldn't work at all as a one-way event source):
- We can expose a list of the files in the current terminal working directory as part of a command, so you can e.g. `git add filename` by saying the filename instead of needing to spell the filename the hard way.
- We can't really create the filename list unless we know you're in a terminal and we know the working directory.
- Ideally we even have an event source telling us when you focus a UI component like a terminal, so we can decide at that point whether to asynchronously go look up filenames and such to prepare for a command.

VSCode is a complex enough application that we can't treat it as just "editor for current language". There are many different UI components that need slightly different treatment, and it's best if the accessibility tool can make decisions about what to do based on the specific kind of UI it's currently automating.

Anecdotally, I know some accessibility users of VSCode who are entirely avoiding the builtin terminal for these reasons.

**Comment by isidorn** [comment]
This issue is better than #118722 since this is one is much more scoped to the actual problem. Exposing `getContext` is too broad for this particular use case.

Using the OS accessibility API approach as proposed by @tyriar sounds reasonable to me. 
Did you maybe try to make a prototype with this approach so we can all try it out and see the limitations? Screen readers use this approach successfully.

**Comment by lunixbochs** [comment]
> this is one is much more scoped to the actual problem

I disagree, this issue is not the core problem, it's just the most annoying instance of it. There's more modal UI in VSCode than the Terminal. I will re-highlight from my previous response:

> VSCode is a complex enough application that we can't treat it as just "editor for current language". There are many different UI components that need slightly different treatment, and it's best if the accessibility tool can make decisions about what to do based on the specific kind of UI it's currently automating.

As for accessibility:

> Using the OS accessibility API approach as proposed by @Tyriar sounds reasonable to me.

I don't think we can rely on accessibility for this unless the performance issues are sorted out, and using the accessibility tree for this is a performance minefield - it's easy to regress. I found this issue about Mac accessibility performance which was closed for inactivity: https://github.com/microsoft/vscode/issues/110764 (It's possibly related to this issue that is still open? https://github.com/microsoft/vscode/issues/41423)

IMO the focused context is such a relatively small piece of information that it's incredibly wasteful to build and repeatedly query a whole-app accessibility tree just to figure it out.

If getContext() is no good because @jrieken wants VSCode's context query parser to own the context information, can we instead solve it like this?
1. Have a `queryContext(...) -> bool` function where we pass in a hotkey-like context and vscode best-effort tells us if it's going to be active right now.
2. Possibly some kind of "some kind of context switch may have just happened" callback into our code?
3. I know that kind of callback could feel expensive to allow, so a way to punt that work to the extension side would be a monotonically increasing `contextSerial() -> int` counter that we can query to see if something may have changed since the last time we evaluated our `queryContext()` list.

It doesn't matter so much if the solution isn't 100% perfect, e.g. has slight race conditions in case of a rapidly updating UI, because accessibility users are far more likely to be in a "paused" state when this needs to be evaluated (like listening to their screen reader, or pausing between voice commands), and are also more likely to be willing to hit undo and keep going than a normal user, because all of the alternatives are worse right now.

**Comment by pokey** [comment]
Any thoughts on this one @Tyriar?  We're happy to help however we can

**Comment by pokey** [comment]
@isidorn this issue means that a critical piece of VSCode functionality is very painful to use for users with accessibility needs.  Can you please comment on why it has been inactive for so long?  Is there something else you're waiting on from us?

**Comment by isidorn** [comment]
We are not awaiting anything from the community. We simply did not have enough time for this request.
It is on the backlog which means it might happen in the future, but no immediate plans. Sorry for not making this clear before.
Thank you for understanding

**Comment by pokey** [comment]
Understood.  If you decide in the future (or now) that you're interested in community contributions here, please let me know.  I work full-time on open-source voice coding software for VSCode, and this issue is a major pain point for our community, so I could certainly justify sinking in some cycles here.  But I get that this touches some fairly core functionality, so can understand the desire to keep it in-house

**Comment by sbungartz** [comment]
Hey there, thanks for the efforts put into this already. I would also love to see some solution for this issue.

I have built a terrible hack right now, and am currently not sure if even I want to keep it because of its side-effects, but I want to at least share the idea:

I built a small plug-in with two commands, one of which adds [TERMINAL] to the window title, and the other removes it. I configured a keybinding for the same key for each command, but with different `when` clauses, so when the terminal has focus the string is added to the title and otherwise it's removed.
The I have a loop (in talon) that presses that key every second if VS Code is open.

This works quite well actually, in terms of always having correct info on terminal focus state, but:
- you cannot really use it with modifier keys, as other keypresses might be randomly combinde with the modifiers
- that means, you need to choose a single key that does not do anything else (I'm currently using "browsersearch")
- switching to another window might cause the keypress to be interpreted in that window
- and most annoying for me right now: The keypress causes hover popups (like for problems and docs) to be closed, drag-selection is interrupted and if you want to use the keyboard holding down arrow keys to navigate also does not really work anymore.

**Comment by meganrogge** [comment]
related https://github.com/microsoft/vscode/issues/97790

**Comment by E1adi** [comment]
Hello everyone,
I also have an extension I'd like to write that requires the functionality described in the opening post of this thread.
I wonder, is it possible to add a property to `TerminalState` such as `isFocused` or something similar?
It will answer both requirements and will be more natural to access.
For the function requirement you can just use the `Terminal.state.isFocused` property.
For the event, we can use `vscode.window.onDidChangeTerminalState`.
Thanks!

**Comment by mqnc** [comment]
@pokey thanks for the hint! I dug a bit deeper and saw that there are quite a lot of issues about this... issue. Is there any progress? I was even considering to evaluate some pixel colors but the only thing that indicates that the terminal has the focus is the blinking cursor. If getContext is not an option, some getActivePanel should surely be possible...

**Comment by sbungartz** [comment]
@mqnc yes just like @pokey said, it only worked with key presses which turned out to have very undesireable effects as listed in my previous post, so I gave it up rather quickly. I am now in a place where I just have the `terminal`-tag and some others always active in `vscode` (see below), and I replaced the prefix `git` with `source` for all the VS-Code-Specific git commands, so I can say `source push` to trigger VS Code's built in push command but `git push` to write that into wherever I am.
This obviously means that sometimes shell commands are written into my code by accident, so the ability to accurately detect terminal focus would still be highly appreciated.

<details>
<summary>Talon Terminal active when in VS Code</summary>

```talon
app: vscode
-
tag(): terminal
tag(): user.generic_unix_shell
tag(): user.git

# Since terminal needs different copy-paste commands than the editor, we need to use this when we are in the terminal:
copy term: key(ctrl-shift-c)
(pace | paste) term: key(ctrl-shift-v)
```

</details>

**Comment by pokey** [comment]
@sbungartz you might try doing insertion via the VSCode command that inserts text in the terminal directly. Then it wouldn't end up in your editor. Can't remember command ID but it's mentioned in VSCode terminal docs

Another possible workaround is to use terminal editors, Ie the ones that appear in an editor group instead of in the bottom panel. Those do show up in the window title

And finally, you can experiment with enabling the full electron accessibility tree, but keep in mind that has performance implications

Fwiw the approach I've taken personally is to add VSCode tasks for my most common terminal commands and map those to voice commands, then just use a keyboard for the rest 😔

**Comment by mrob95** [comment]
May be an obvious point but you can also just alt tab to a separate terminal program. I don't find it too much of a burden to switch between them as necessary - it's one voice command whether you are switching within VSCode or switching to a different window.

**Comment by pokey** [comment]
@Tyriar @meganrogge fwiw even just having a way to make it so that the window title reflects whether the terminal is focused would be great; arguably that would be even easier to leverage, as talon has really good support for watching the window title. I believe there's a separate issue for that one

**Comment by Tyriar** [comment]
@pokey I guess that ask is for a new focused view variable in `window.title`? That sounds like it would be fairly simple and uncontroversial

**Comment by pokey** [comment]
That would be amazing yeah. I believe it's tracked by this issue fwiw https://github.com/microsoft/vscode/issues/97790

**Comment by pokey** [comment]
Yeah would also be useful if it could indicate that an editor has focus, as often we want to know "am I typing into an editor or not?". If the answer is "yes" we can be slick, eg use snippets, but fall back to simulating keypresses when the answer is "no"

**Comment by mqnc** [comment]
> @Tyriar @meganrogge fwiw even just having a way to make it so that the window title reflects whether the terminal is focused would be great; arguably that would be even easier to leverage, as talon has really good support for watching the window title. I believe there's a separate issue for that one

What has to happen so this gets implemented? Is it worth looking into this myself and making a PR? (I have never looked at the vscode source but I have some experience with JS and TS).

**Comment by meganrogge** [comment]
I am not sure what would be needed here, so you might have to poke around yourself. It might be as simple as displaying the context key of `terminalFocus` in a new variable as mentioned above

> I guess that ask is for a new "focused view" variable in window.title? That sounds like it would be fairly simple and uncontroversial

**Comment by Tyriar** [comment]
@mqnc implementing this is quite easy, it needs a champion on the team to go to the API sync (again) to advocate and talk through what the actual API will be.

**Comment by meganrogge** [comment]
Now that we have #97790, is there a use case for this API?

**Comment by meganrogge** [comment]
@pokey pls lmk if you still think we need to do this

**Comment by VSCodeTriageBot** [comment]
This issue has been closed automatically because it needs more information and has not had recent activity. See also our [issue reporting](https://aka.ms/vscodeissuereporting) guidelines.

Happy Coding!

**Comment by pokey** [comment]
> @pokey pls lmk if you still think we need to do this

Yeah agreed let's park this one for now; it's not urgently necessary for us. The window var PRs were huge 🙌

**Comment by ClemensDinkel** [comment]
With [#97790](https://github.com/microsoft/vscode/issues/97790) resolved, we still don't have have access to the current focus programmatically, do we ? 
I'm currently trying to write an extension to insert a string into the focused text editor or terminal, depending on what's currently focused. For simplicity I wan't both operations to run on the same command. In the when clause I can simply say ` "when": "editorTextFocus || terminalFocus"` but in my extension code, when I'm registering the command I have no way of knowing which one is currently focused. `vscode.window.activeTextEditor` and `vscode.window.activeTerminal` are both defined, since they return the last active instance, even if they're not currently in focus.

**Comment by ClemensDinkel** [comment]
Yes that's what I'm doing atm and they share the same shortcut, since they will never activate both at the same time. This gets close, but I still need to list seperate commands which feels a bit unsatisfying. Not a big issue, but still wanted to mention there is use cases for this scenario.

**Comment by valerybugakov** [comment]
Hey folks, do you have any updates on this? We've got a use case for this API. We want to process text document edits that users make (`onDidChangeTextDocument`), but only when the editor is focused. This would help us avoid handling changes triggered by version control systems and other third-party tools that might be activated from the terminal.
